metadata-ingestion/src/datahub/ingestion/source/nifi.py

```diff
@@ -360,7 +428,9 @@ class NifiFlow:
     root_process_group: NifiProcessGroup
     components: Dict[str, NifiComponent] = field(default_factory=dict)
     remotely_accessible_ports: Dict[str, NifiComponent] = field(default_factory=dict)
-    connections: List[Tuple[str, str]] = field(default_factory=list)
+    connections: BidirectionalComponentGraph = field(
+        default_factory=BidirectionalComponentGraph
+    )
```

Comment:
Assuming, perf improvement is due to removing duplicate connection edges, 
I wonder if changing `connections: List[Tuple[str, str]]`  to `connections: Set[Tuple[str, str]]` and keeping everything else just as before give similar boost in perf ?

Reply:
It would address duplicates problem but then we would need to still iterate over whole set to find matches for incoming/outgoing connections for an entity. The question is why would we do it this way then if we already have better implementation in place? Do you see any shortcoming of the current implementation, beside taking roughly twice the amount of memory (which I think is not much in first place, amount of connections, even for the biggest clusters, tends to be below 100k).

Reply:
The implementation of `BidirectionalComponentGraph` is fair and is adapted as per operations we need. It looks good to me.